File: solutions/0156_Binary_Tree_Upside_Down/dfs.java
// --Link--
// https://www.lintcode.com/problem/binary-tree-upside-down/description

/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */

public class Solution {
    /**
     * @param root: the root of binary tree
     * @return: new root
     */
    public TreeNode upsideDownBinaryTree(TreeNode root) {
        // write your code here
        return flip(root);
    }

    public TreeNode flip(TreeNode node) {
        if (node == null) return null;
        if (node.left == null && node.right == null) return node;
        TreeNode l = flip(node.left);
        TreeNode r = node.right; // maybe null
        // find l's rightmost successor
        TreeNode runner = l;
        while (runner.right != null) runner = runner.right;

        node.left = null;
        node.right = null;
        runner.left = r;
        runner.right = node;
        return l;
    }
}

File: solutions/0543_Diameter_of_Binary_Tree/dfs.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:

    def diameterOfBinaryTree(self, root: TreeNode) -> int:
        self.ret = 0
        self.longestPathFromNode(root)
        return self.ret

    def longestPathFromNode(self, node):
        if not node.left and not node.right:
            return 0
        l = 0 if not node.left else (1 + self.longestPathFromNode(node.left))
        r = 0 if not node.right else (1 + self.longestPathFromNode(node.right))
        self.ret = max(l + r, self.ret)
        return max(l, r)
